src/core/components/block.js

```diff
@@ -81,17 +81,19 @@ module.exports = function block (self) {
             cb(null, new Block(block, cid))
           })
         },
-        (block, cb) => self._blockService.put(block, (err) => {
-          if (err) {
-            return cb(err)
-          }
+        (block, cb) => self._gcLock.readLock((_cb) => {
```

Comment:
I think we don't want to write to the blockstore while GC is running, because depending on where GC is up to while iterating through the blocks, the block may or may not be GCed, so it results in non-deterministic behaviour.